torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,395 @@
+  // corresponding value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
+      Node* node,
+      std::string& name,
+      script::Module& attrModule) {
+    names_.clear();
+    while (!(node->outputs()[0]->type() == attrModule.type())) {
+      if (node->kind() == prim::GetAttr) {
+        names_.push_front(node->s(attr::name));
+        node = node->inputs()[0]->node();
+      } else {
+        return false;
+      }
+    }
+
+    for (auto& moduleName : names_) {
+      auto it = preservedAttrs_.find(attrModule._ivalue());
+      if (it != preservedAttrs_.end()) {
+        if (it->second.count(attrModule.attr(moduleName))) {
+          return false;
+        }
+      }
+      attrModule = attrModule.attr(moduleName).toModule();
+    }
+
+    auto attr = attrModule.attr(name);
+    if (!AliasDb::mutableType(attr.type())) {
+      auto it = preservedScalarAttrs_.find(attrModule._ivalue());
+      return it == preservedScalarAttrs_.end() || !it->second.count(name);
+    }
+
+    auto it = preservedAttrs_.find(attrModule._ivalue());
+    if (it == preservedAttrs_.end()) {
+      return true;
+    }
+    if (it->second.count(attr)) {
+      return false;
+    }
+    if (!attr.type()->cast<ClassType>()) {
+      for (auto& ivalue : it->second) {
+        if (!ivalue.isObject() && ivalue.overlaps(attr)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  void insertMutableAttr(
+      const std::string& name,
+      const IValue& attr,
+      script::Module& attrModule) {
+    if (AliasDb::mutableType(attr.type())) {
+      preservedAttrs_[attrModule._ivalue()].insert(attr);
+    } else {
+      preservedScalarAttrs_[attrModule._ivalue()].insert(name);
+    }
+  }
+
+  void recordMutableAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    std::unique_ptr<AliasDb> aliasDb =
+        torch::make_unique<AliasDb>(graph, /* isFrozen */ true);
+    IValue::HashAliasedIValues usedAttrs;
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr || n->kind() == prim::GetAttr) {
+          // TODO: handle interface attributes. For now, Exit if Module uses
+          // inteface attributes
+          if (n->kind() == prim::GetAttr) {
+            TORCH_CHECK(
+                !n->output()->type()->cast<InterfaceType>(),
+                "attempted to freeze a module that uses interface attributes");
+          }
+          auto inputNode = n->inputs()[0]->node();
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
```

Comment:
It's a little confusing that we're calling `findConstantAttr`, in the function that sets up the correctness conditions for `findConstantAttr`. Maybe add comment. Are we doing this to not re-traverse mutable attributes ?